src/test/java/com/microsoft/azure/relay/AsyncLockTest.java

```diff
@@ -119,4 +121,91 @@ public void lockTimeoutTest() throws InterruptedException, ExecutionException, T
 		LockRelease lockRelease3 = lockFuture3.get(2000, TimeUnit.MILLISECONDS);
 		lockRelease3.release();
 	}
+	
+	@Test
+	public void lockScopeUncontended() throws Exception {
+		String originalString = "lockScopeUncontended";
+		AsyncLock lock = new AsyncLock();
+		assertFalse("Should not be locked to start", lock.isLocked());
+		
+		CompletableFuture<String> innerTask = new CompletableFuture<String>();
+		CompletableFuture<String> outerTask = lock.runInsideLockAsync(null, EXECUTOR, () -> {
+			assertTrue("Should be locked during Inner Function", lock.isLocked());
+			return innerTask;
+		});
+		
+		assertFalse(outerTask.isDone());
+		assertTrue("Should be locked now", lock.isLocked());
+		innerTask.complete(originalString);
```

Comment:
>innerTask.complete(originalString); [](start = 2, length = 35)

Should we check if outerTask is complete here before we manually call it to complete?

Reply:
line 137 asserts that outerTask is not done. Is that what you're looking for? Or do you want the test to wait a little bit like this:

  Throwable error = assertThrows(ExecutionException.class, (Executable)() -> outerTask.get(TIMEOUT_MS, TimeUnit.MILLISECONDS));
  assertEquals(TimeoutException.class, error.getCause().getClass());

---
In reply to: [259048928](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259048928) [](ancestors = 259048928)

Reply:
I meant like outerTask should be automatically completed right after we call innerTask.complete(). But that's beside the point of the test I guess

---
In reply to: [259059280](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259059280) [](ancestors = 259059280,259048928)

Reply:
Isn't what you're describing exactly what the test is doing? After calling innerTask.complete the next line gives outerTask TIMEOUT_MS to complete otherwise get throws and the test fails.

---
In reply to: [259060993](https://github.com/Azure/azure-relay-java/pull/20#discussion_r259060993) [](ancestors = 259060993,259059280,259048928)